### ISSUE TITLE
[feature] Support offline tokenized + index-only packed mode for SFT datasets

### ISSUE THREAD

**Issue by muyihao** [NVIDIA-NeMo/Megatron-Bridge#4351]
### User problem

  Megatron-Bridge currently supports SFT sequence packing through `FinetuningDatasetBuilder` and `prepare_packed_sequence_data()`.

  The current flow prepares packed data before training by:

  1. loading the raw SFT dataset;
  2. tokenizing examples through the SFT dataset path;
  3. building a first-fit packing strategy;
  4. materializing packed rows to disk.

  The materialized packed format currently stores packed token rows, for example:

  - `input_ids`
  - `loss_mask`
  - `seq_start_id`

  in packed parquet format, with `.npy` also supported as a legacy format.

  This works well for normal SFT use cases, but for large-scale SFT / mid-training / CPT-style workloads, materializing every packed token row has several limitations:

  - the packed output duplicates token data that may already exist in an offline tokenized dataset;
  - changing `packed_sequence_size`, packing seed, or packing algorithm requires regenerating large packed token files;
  - it is harder to reuse the same tokenized corpus across different packing strategies;

### Desired outcome

Add an `index_only` output mode for SFT sequence packing, splitting preparation into two reusable stages plus a lazy training-time reader.

  #### Stage 1: offline tokenization

  Add or expose a preprocessing path such as:

  ```python
  prepare_tokenized_data(...)
  ```

  This stage writes a reusable tokenized dataset, for example in:

  - a Bridge-native mmap format;
  - an AIRecord-compatible tokenized format;
  - or another random-access tokenized sample store.

  Each tokenized sample would contain the fields needed by SFT training, such as:

  - `input_ids`
  - `loss_mask`
  - optional metadata required for chat / answer-only-loss semantics

  #### Stage 2: index-only packing

  Add a packing preparation path such as:

  ```python
  prepare_packed_index_data(...)
  ```

  When `output_format="index_only"` is enabled, this stage should only write packing metadata / indices instead of materialized packed token rows. For each packed sample, the index could store entries like:

  ```text
  pack_id -> [
    (doc_index, start_offset, end_offset),
    (doc_index, start_offset, end_offset),
    ...
  ]
  ```

  The packing index should preserve the existing semantics of:

  - `packed_sequence_size`
  - `first_fit_shuffle`
  - `first_fit_decreasing`
  - deterministic seed behavior
  - `seq_start_id` / `cu_seqlens` construction
  - `loss_mask` alignment
  - `pad_seq_to_mult`
  - `pad_cu_seqlens` where applicable

  #### Training-time reader

  Add a lazy packed dataset reader, for example `PackedIndexSFTDataset`, or extend the existing packed dataset path with something like `lazy_load=True` / `output_format="index_only"`. At training time, the reader
  would:

  1. read one packed-index row;
  2. fetch the referenced tokenized samples from the underlying tokenized store;
  3. concatenate `input_ids` and `loss_mask` dynamically;
  4. construct `seq_boundaries`, `position_ids`, `cu_seqlens`, and labels using the existing packed SFT semantics.

  #### Expected benefits

  - avoiding duplicated materialized packed token data;
  - allowing the same tokenized dataset to be reused across different packing configurations;
  - reducing preprocessing output size;
  - making packing regeneration faster when only packing indices change;
  - aligning Bridge SFT data preparation with offline tokenized / packed-index data pipelines;

### Alternatives considered

  - Keep the current materialized packed parquet / `.npy` flow and regenerate packed files whenever `packed_sequence_size`, packing seed, or packing algorithm changes. Workable for small datasets, but expensive in
  storage and preprocessing time at large scale.
  - Pre-tokenize externally and feed token-level data through the existing SFT path, but this still forces re-materialization of packed token rows for every packing configuration and does not let multiple packing
  strategies share one tokenized corpus.
  - Implement packing entirely on the fly at training time without an index, which loses the determinism and reproducibility guarantees of the current first-fit packing strategies (`first_fit_shuffle`,
  `first_fit_decreasing`, seeded behavior).

### Affected area

area:data

### Urgency / use case

Blocking current work

**Comment by yaoyu-33** [comment]
Hi, it's a valid feature ask. We need a bit more time to work on this.

**Comment by cuichenx** [comment]
> Implement packing entirely on the fly at training time without an index, which loses the determinism and reproducibility guarantees of the current first-fit packing strategies (first_fit_shuffle,
first_fit_decreasing, seeded behavior).
Could you explain why it loses determinism? On-the-fly packing will produce different packs but it should still be deterministic within its own framework. We have on-the-fly packing for VLM datasets currently and we plan on bringing it to LLMs in the future. Does that work for your use case? 

If I understand correctly, in your proposal, a change in the model/tokenizer will still require repacking offline, right? So the win only materializes when you sweep packing configs on a fixed model/tokenizer?

**Comment by muyihao** [comment]
@cuichenx Thanks for the follow-up! On determinism — you're right, I overstated it. Seeded on-the-fly packing is deterministic. The main benefit is when the model and tokenizer remain unchanged. In that case, if you only want to experiment with different packing strategies or configurations, you can reuse the already tokenized data instead of re-running the tokenization step.

In our experience, tokenization is by far the most time-consuming stage, typically taking hours for large datasets, while generating new packing results from the tokenized data usually takes only a few minutes.

**Comment by cuichenx** [comment]
Thanks for the clarification. We have had similar observations, and offline tokenization is a planned feature (it will use the same .bin .idx format as megatron pretraining). That combined with either offline or online packing should mitigate most of the time spent in data preparation.

**Comment by muyihao** [comment]
@cuichenx We actually have an internal implementation based on Megatron's .bin/.idx indexed dataset format. It performs offline tokenization once, allowing the tokenized data to be reused across multiple training runs with different packing strategies.

In addition, it supports packing across multiple data sources, which can be naturally combined with Megatron's dataset blending strategy. We've found this to be quite practical for real-world training workflows. The implementation currently supports both pretraining and SFT datasets.

If this aligns with the direction you're planning for Megatron-Bridge, we'd be happy to contribute it back to the project. 🙂

**Comment by liding-nv** [comment]
@cuichenx checking if we have agreed on a plan to move forward?

**Comment by huvunvidia** [comment]
Hi @cuichenx ,
Seems that this direction align with M-Bridge's plan.
How do you prefer to move forward? We can have @ muyihao contribute PRs and you can review.

**Comment by cuichenx** [comment]
Sorry for the delay. Please check tracking issue https://github.com/NVIDIA-NeMo/Megatron-Bridge/issues/4596
Does https://github.com/NVIDIA/Megatron-LM/pull/5017 fit what you have in mind? @muyihao

**Comment by muyihao** [comment]
@cuichenx Thanks for the pointer! Yes, at a high level, #5017 looks aligned with what I had in mind and should cover the core requirement. I’ll take a closer look at the implementation details and follow up if I find any gaps or areas where I can help.

**Comment by bzantium** [comment]
This one may be blocked on a status question. NVIDIA-NeMo/Megatron-Bridge#4596, which this issue was redirected to, was closed on 2026-08-07, but the upstream pieces still look open: NVIDIA/Megatron-LM#5017 and NVIDIA/Megatron-LM#5409 are unmerged, `megatron/core/datasets/sft_dataset.py` is not on Megatron-LM `main`, and Bridge does not consume an MCore SFT dataset yet. I asked there for clarification and am repeating the question here since this thread has been waiting on the same answer.

- Is the pretokenized `.bin`/`.idx` SFT path still the intended direction, and is it deferred rather than dropped?
- Should packed Parquet be treated as the supported format in the meantime?

There is also a smaller gap that is independent of how that lands. Offline packing is available on the materialized `GPTSFTDatasetConfig` path but not on the direct Hugging Face path, so a direct-HF run cannot use packed sequences at all today. The existing packing algorithms already work on tokenized rows rather than files, so closing that gap needs an adapter rather than a second packing implementation. I would be glad to contribute it if that is welcome.
